File: go/cargo/logger.go
package cargo

import (
	"log"
	"os"
)

// Logger is the minimal logging interface the cargo package uses.
// Implementations are responsible for prefixes, timestamps, output etc.
type Logger interface {
	Infof(format string, args ...any)
	Warnf(format string, args ...any)
}

// infoPrefix / warnPrefix are computed once at startup based on whether
// stderr is a TTY and whether NO_COLOR is set
// (https://no-color.org/ -- presence of the env var disables colour
// regardless of value).
var (
	infoPrefix string
	warnPrefix string
)

func init() {
	if colorEnabled() {
		infoPrefix = "\033[32m[INFO]\033[0m "
		warnPrefix = "\033[33m[WARN]\033[0m "
	} else {
		infoPrefix = ""
		warnPrefix = "warning: "
	}
}

func colorEnabled() bool {
	if _, set := os.LookupEnv("NO_COLOR"); set {
		return false
	}
	fi, err := os.Stderr.Stat()
	if err != nil {
		return false
	}
	return (fi.Mode() & os.ModeCharDevice) != 0
}

type stdLogger struct{}

func (stdLogger) Infof(format string, args ...any) { log.Printf(infoPrefix+format, args...) }
func (stdLogger) Warnf(format string, args ...any) { log.Printf(warnPrefix+format, args...) }

File: rust/src/logger.rs
use std::io::IsTerminal;

#[derive(Clone)]
pub struct Logger {
    info_prefix: &'static str,
    warn_prefix: &'static str,
}

impl Logger {
    pub fn new() -> Self {
        let color = color_enabled();
        Self {
            info_prefix: if color { "\x1b[32m[INFO]\x1b[0m " } else { "" },
            warn_prefix: if color {
                "\x1b[33m[WARN]\x1b[0m "
            } else {
                "warning: "
            },
        }
    }

    pub fn info(&self, msg: &str) {
        eprintln!("{}{}", self.info_prefix, msg);
    }

    pub fn warn(&self, msg: &str) {
        eprintln!("{}{}", self.warn_prefix, msg);
    }
}

impl Default for Logger {
    fn default() -> Self {
        Self::new()
    }
}

fn color_enabled() -> bool {
    if std::env::var("NO_COLOR").is_ok() {
        return false;
    }
    std::io::stderr().is_terminal()
}
